packages/material-ui/src/Fade/Fade.test.js

```diff
@@ -28,95 +28,106 @@ describe('<Fade />', () => {
-              assert.strictEqual(handleEntering.callCount, 1);
-              assert.strictEqual(handleEntering.args[0].length, 2);
-              done();
-            }}
-            {...defaultProps}
-          >
-            <div />
-          </Fade>,
-        );
-      });
+  describe('transition lifecycle', () => {
+    let clock;
+    let wrapper;
+    let child;
+
+    const handleEnter = spy();
+    const handleEntering = spy();
+    const handleEntered = spy();
+    const handleExit = spy();
+    const handleExiting = spy();
+    const handleExited = spy();
+
+    before(() => {
+      clock = useFakeTimers();
+      wrapper = mount(
+        <Fade
+          onEnter={handleEnter}
+          onEntering={handleEntering}
+          onEntered={handleEntered}
+          onExit={handleExit}
+          onExiting={handleExiting}
+          onExited={handleExited}
+        >
+          <div id="test" />
+        </Fade>,
+      );
+      child = wrapper.find('#test');
+    });
+
+    after(() => {
+      clock.restore();
     });
 
-    describe('exiting', () => {
-      it('should fire callbacks', done => {
-        const handleExit = spy();
-        const handleExiting = spy();
-
-        const wrapper = mount(
-          <Fade
-            onExit={handleExit}
-            onExiting={handleExiting}
-            onExited={() => {
-              assert.strictEqual(handleExit.callCount, 1);
-              assert.strictEqual(handleExit.args[0].length, 1);
-              assert.strictEqual(handleExiting.callCount, 1);
-              assert.strictEqual(handleExiting.args[0].length, 1);
-              done();
-            }}
-            {...defaultProps}
-          >
-            <div />
-          </Fade>,
-        );
-
-        wrapper.setProps({
-          in: false,
+    describe('in', () => {
+      before(() => {
+        wrapper.setProps({ in: true });
+      });
+
+      describe('handleEnter()', () => {
+        it('should call handleEnter()', () => {
+          assert.strictEqual(handleEnter.callCount, 1);
+          assert.strictEqual(handleEnter.args[0][0], child.instance());
+        });
+
+        it('should set style properties', () => {
+          assert.match(
+            handleEnter.args[0][0].style.transition,
+            /opacity 225ms cubic-bezier\(0.4, 0, 0.2, 1\)( 0ms)?/,
+          );
         });
       });
-    });
-  });
 
-  describe('transition lifecycle', () => {
-    describe('handleEnter()', () => {
-      it('should set style properties', () => {
-        mount(
-          <Fade
-            {...defaultProps}
-            onEnter={node => {
-              // Needs to be match because Edge doesn't include the 0ms delay
-              assert.match(
-                node.style.transition,
-                /opacity 225ms cubic-bezier\(0.4, 0, 0.2, 1\)( 0ms)?/,
-              );
-            }}
-          />,
-        );
+      describe('handleEntering()', () => {
+        it('should call handleEntering()', () => {
```

Comment:
My inner demon wants to complain about using proper matchers again. We could make a stateless test suite while using proper descriptions. This is why `spy.should.always.have.been.calledOnceWith(child.instance())` or `assert.calledOnceWith(spy, child.instance())` exist (both need setup).

Now that this itch is satisfied I'm satisfied too :smile:

Reply:
@eps1lon `it.shall.not.pass`

I can find the usage of `.calledWith()` and `.calledWithMatch()` in the codebase. It uses *Sinon* to generate a boolean. We might as well use `.args()[x][y]`. I don't see a clear benefit in this approach. It still using assert.strictEqual. What I see is the disadvantage of the two solutions. assert.strictEqual fail messages are cryptic, they can be slow to debug and `i.can.write.test.in.english` is very hard to guess for a non initiated user.

I assume that the whole point of `assert.calledOnceWith()` is to improve the fail message. I have no objection with a cherry pick of the best test assertion helpers. The one that are common enough to be found by a new user and useful enough to save time when the test fail. What I want to avoid is a systemic approach where we write a soup of english API that nobody else can guess.

Reply:
> i.can.write.test.in.english is very hard to guess for a non initiated user.

You want to tell me that natural language is harder than a programming language?

> What I want to avoid is a systemic approach where we write a soup of English APIs that nobody else can guess.

Who is nobody? Absolute beginners? People that can speak a programming language but not english? Because it's definitely not people familiar with any popular testing framework.

>  It uses Sinon to generate a boolean. We might as well use .args()[x][y]

We're not having an honest discussion here. Preferring our own API with call + nested property access over plain english is just gatekeeping in its purest form.

Reply:
> You want to tell me that natural language is harder than a programming language?

Yes!

Reply:
jest and chai use these matchers. I would bet that any assert/expect library uses these matchers at which point I'm inclined to ask: What do you know that nobody else seems to get? Especially the part about a programming language being better than natural language i.e. better be as far away from natural language as possible I have never heard. Any literature on that?

Reply:
Jest and Chai API is different. You have to learn them.